### PR TITLE
rand for GL and Heisenberg group

### PR DESCRIPTION
**Pull request by mateuszbaran**
Useful for testing here: #609 .

### CODE REVIEW

File: src/groups/general_linear.jl [hunk 1]
```diff
@@ -246,6 +246,15 @@ project(::GeneralLinear, p, X) = X
 project!(::GeneralLinear, q, p) = copyto!(q, p)
 project!(::GeneralLinear, Y, p, X) = copyto!(Y, X)
 
+function Random.rand!(G::GeneralLinear, pX; kwargs...)
```
Comment (kellertuer) [inline src/groups/general_linear.jl:249]: Would that maybe be a generic good idea for certain manifolds (isometrically embedded ones or such?) to pass rand to the embedding?

**Review by kellertuer — CHANGES_REQUESTED** [review]
My only remark for this PR is (the typical Ronny-one), that both new rand methods should be documented.

File: src/groups/general_linear.jl (hunk 1 above)
Comment (mateuszbaran) [reply src/groups/general_linear.jl:249]: I would prefer to avoid trait dispatch here to keep the code simple.

Comment (kellertuer) [reply src/groups/general_linear.jl:249]: Sure, that is also fine :)

**Comment by mateuszbaran** [comment]
I was actually somewhat expecting another long discussion whether this is the right distribution :sweat_smile:

File: src/groups/general_linear.jl [hunk 2]
```diff
@@ -246,6 +246,27 @@ project(::GeneralLinear, p, X) = X
 project!(::GeneralLinear, q, p) = copyto!(q, p)
 project!(::GeneralLinear, Y, p, X) = copyto!(Y, X)
 
+
+@doc raw"""
+    Random.rand(G::GeneralLinear; kwargs...0)
```
Comment (kellertuer) [inline src/groups/general_linear.jl:251]:
```suggestion
    Random.rand(G::GeneralLinear; vector_at=nothing, kwargs...)
```

This way the default is mentioned (and I removed a typo)

**Review by kellertuer — APPROVED** [review]
Besides the one typo, this looks good to me.

File: src/groups/semidirect_product_group.jl [hunk 3]
```diff
@@ -123,13 +123,15 @@ function _compose!(G::SemidirectProductGroup, x, p, q)
...
+    x_tmp = allocate(x)
     np, hp = submanifold_components(G, p)
     nq, hq = submanifold_components(G, q)
-    nx, hx = submanifold_components(G, x)
+    nx, hx = submanifold_components(G, x_tmp)
     compose!(H, hx, hp, hq)
     nxtmp = apply(A, hp, nq)
     compose!(N, nx, np, nxtmp)
     @inbounds _padpoint!(G, x)
+    copyto!(x, x_tmp)
```
Comment (mateuszbaran) [inline src/groups/semidirect_product_group.jl:134]: This is the bugfix from https://github.com/JuliaManifolds/Manifolds.jl/issues/609 -- I just didn't want to open yet another PR...

Comment (mateuszbaran) [reply src/groups/semidirect_product_group.jl:134]: Basically `p` and `x` were the same variable and overwriting `x` caused `p` to change.

Comment (kellertuer) [reply src/groups/semidirect_product_group.jl:134]: Oh, indeed that was a side effect that should not happen, well-spotted. Sure we can add it here as well. Though at some point we should maybe also start a Changelog here (to see where we did what also here).

Comment (mateuszbaran) [reply src/groups/semidirect_product_group.jl:134]: Yes, I would really like to have a changelog here but I'm usually less enthusiastic about actually writing one :sweat_smile:

Comment (kellertuer) [reply src/groups/semidirect_product_group.jl:134]: Then let's start one and I take the role (similar to the docs) of reminding everyone (hopefully, forgot that twice in Manopt already as well and did one PR with 3 changelog entries).

I think it's worth it, especially for the really condensed overview of when which topic/feature appeared.

**Pull request merged**